usr/src/uts/common/brand/lx/syscall/lx_socket.c

```diff
@@ -240,7 +242,25 @@ static const int stol_socktype[SOCK_SEQPACKET + 1] = {
 #define	STOL_SOCKTYPE(t)	\
 	((t) <= SOCK_SEQPACKET ? stol_socktype[(t)] : SOCK_INVAL)
 
+/* illumos represents tcp states with ranges between -6 to 6, where:
+ * tcp states > 0 define connections that have been stablished.
+ * tcp states < 0 define connections that have not yet stablished.
+ * Where Linux defines their tcp states as a range from 1 to 12.
+ * the macro STOL_TCPSTATE translate our illumos tcp states to a Linux one,
+ * in case there is no tcp state returned by OS that matches Linux, the state
+ * LX_TCP_CLOSE_WAIT will be returned.
+ * */
+static const int stol_tcp_state[LX_TCP_NEW_SYN_RECV + 1] = {
+	LX_TCP_ESTABLISHED, LX_TCP_CLOSE_WAIT, LX_TCP_FIN_WAIT1,
+	LX_TCP_CLOSING, LX_TCP_LAST_ACK, LX_TCP_FIN_WAIT2, LX_TCP_TIME_WAIT,
+	LX_TCP_SYN_RECV, LX_TCP_SYN_SENT, LX_TCP_LISTEN, LX_TCP_NEW_SYN_RECV,
+	LX_TCP_CLOSE_WAIT, LX_TCP_CLOSE_WAIT
+};
 
+#define STOL_TCPSTATE(t)        \
+        (((t) <= LX_OS_MAX_TCP_STATE && (t) >= LX_OS_MIN_TCP_STATE )? \
+		((t) < 0 ? stol_tcp_state[((t) * -1)  + 6] : \
```

Comment:
Seems like a lot of extra math for something that, since you *are* using a table, could be something like:

```
#define STOL_TCPSTATE(t) \
        (((t) <= LX_OS_MAX_TCP_STATE && (t) >= LX_OS_MIN_TCP_STATE )? \
                stol_tcp_state[(t) + 6] : LX_TCP_CLOSE_WAIT)
```
As long as your define your table right.  I see Linux has no TCPS_IDLE equivalent so THAT's why you return CLOSE_WAIT.  Can you literally run a test on Linux where you just s = socket(AF_INET, SOCK_STREAM, 0); and run TCP_INFO on it to see what you get there?

Reply:
For the macro I think the extra math is required,  the array is partitioned between the available values for illumos TCP states, and negative values from illumos are offset to fall in the negative range of the table. 
I have a simple test program to check this:
```c
#include<stdio.h>


#define LX_TCP_ESTABLISHED  1
#define LX_TCP_SYN_SENT  2
#define LX_TCP_SYN_RECV  3
#define LX_TCP_FIN_WAIT1  4
#define LX_TCP_FIN_WAIT2  5
#define LX_TCP_TIME_WAIT  6
#define LX_TCP_CLOSE  7
#define LX_TCP_CLOSE_WAIT  8
#define LX_TCP_LAST_ACK  9
#define LX_TCP_LISTEN  10
#define LX_TCP_CLOSING  11
#define LX_TCP_NEW_SYN_RECV  12
#define LX_OS_MAX_TCP_STATE 6
#define LX_OS_MIN_TCP_STATE -6


/* TCP states */
#define TCPS_CLOSED             -6
#define TCPS_IDLE               -5      /* idle (opened, but not bound) */
#define TCPS_BOUND              -4      /* bound, ready to connect or accept */
#define TCPS_LISTEN             -3      /* listening for connection */
#define TCPS_SYN_SENT           -2      /* active, have sent syn */
#define TCPS_SYN_RCVD           -1      /* have received syn (and sent ours) */
/* states < TCPS_ESTABLISHED are those where connections not established */
#define TCPS_ESTABLISHED        0       /* established */
#define TCPS_CLOSE_WAIT         1       /* rcvd fin, waiting for close */
/* states > TCPS_CLOSE_WAIT are those where user has closed */
#define TCPS_FIN_WAIT_1         2       /* have closed and sent fin */
#define TCPS_CLOSING            3       /* closed, xchd FIN, await FIN ACK */
#define TCPS_LAST_ACK           4       /* had fin and close; await FIN ACK */
/* states > TCPS_CLOSE_WAIT && < TCPS_FIN_WAIT_2 await ACK of FIN */
#define TCPS_FIN_WAIT_2         5       /* have closed, fin is acked */
#define TCPS_TIME_WAIT          6       /* in 2*msl quiet wait after close */


static const int stol_tcp_state[LX_TCP_NEW_SYN_RECV + 1] = {
        LX_TCP_ESTABLISHED,
        LX_TCP_CLOSE_WAIT,
        LX_TCP_FIN_WAIT1,
        LX_TCP_CLOSING,
        LX_TCP_LAST_ACK,
        LX_TCP_FIN_WAIT2,
        LX_TCP_TIME_WAIT, // -- 6
        LX_TCP_SYN_RECV, // -1
        LX_TCP_SYN_SENT, // - 2
        LX_TCP_LISTEN, // -3
        LX_TCP_CLOSE_WAIT, // -4
        LX_TCP_CLOSE_WAIT,  // -5
        LX_TCP_CLOSE // - 6
};

#define STOL_TCPSTATE(t)        \
        (((t) <= LX_OS_MAX_TCP_STATE && (t) >= LX_OS_MIN_TCP_STATE )? \
                ((t) < 0 ? stol_tcp_state[((t) * -1)  + 6] : \
                        stol_tcp_state[(t)]) : LX_TCP_CLOSE_WAIT)
int
main ()
{

  printf ("TCPS_ESTABLISHED %d LX_ESTABLISHED %d result %d---> %s \n",
          TCPS_ESTABLISHED, LX_TCP_ESTABLISHED, STOL_TCPSTATE (0),
          STOL_TCPSTATE (0) == LX_TCP_ESTABLISHED ? "OK" : "ERR");

  printf ("TCPS_CLOSE_WAIT %d LX_CLOSE_WAIT %d result %d---> %s \n",
          TCPS_CLOSE_WAIT, LX_TCP_CLOSE_WAIT, STOL_TCPSTATE (1),
          STOL_TCPSTATE (1) == LX_TCP_CLOSE_WAIT ? "OK" : "ERR");

  printf ("TCPS_FIN_WAIT_1 %d LX_TCP_FIN_WAIT1 %d result %d---> %s \n",
          TCPS_FIN_WAIT_1, LX_TCP_FIN_WAIT1, STOL_TCPSTATE (2),
          STOL_TCPSTATE (2) == LX_TCP_FIN_WAIT1 ? "OK" : "ERR");

  printf ("TCPS_CLOSING %d LX_TCP_CLOSING %d result %d---> %s \n",
          TCPS_CLOSING, LX_TCP_CLOSING, STOL_TCPSTATE (3),
          STOL_TCPSTATE (3) == LX_TCP_CLOSING ? "OK" : "ERR");

  printf ("TCPS_LAST_ACK %d LX_TCP_LAST_ACK %d result %d---> %s \n",
          TCPS_LAST_ACK, LX_TCP_LAST_ACK, STOL_TCPSTATE (4),
          STOL_TCPSTATE (4) == LX_TCP_LAST_ACK ? "OK" : "ERR");

  printf ("TCPS_FIN_WAIT_2 %d LX_TCP_FIN_WAIT2 %d result %d---> %s \n",
          TCPS_FIN_WAIT_2, LX_TCP_FIN_WAIT2, STOL_TCPSTATE (5),
          STOL_TCPSTATE (5) == LX_TCP_FIN_WAIT2 ? "OK" : "ERR");


  printf ("TCPS_TIME_WAIT %d LX_TCP_TIME_WAIT %d result %d---> %s \n",
          TCPS_TIME_WAIT, LX_TCP_TIME_WAIT, STOL_TCPSTATE (6),
          STOL_TCPSTATE (6) == LX_TCP_TIME_WAIT ? "OK" : "ERR");

  printf ("TCPS_SYN_RCVD %d LX_TCP_SYN_RECV %d result %d---> %s \n",
          TCPS_SYN_RCVD, LX_TCP_SYN_RECV, STOL_TCPSTATE (-1),
          STOL_TCPSTATE (-1) == LX_TCP_SYN_RECV ? "OK" : "ERR");

  printf ("TCPS_SYN_SENT %d LX_TCP_SYN_SENT %d result %d---> %s \n",
          TCPS_SYN_SENT, LX_TCP_SYN_SENT, STOL_TCPSTATE (-2),
          STOL_TCPSTATE (-2) == LX_TCP_SYN_SENT ? "OK" : "ERR");

  printf ("TCPS_LISTEN %d LX_TCP_LISTEN %d result %d---> %s \n",
          TCPS_LISTEN, LX_TCP_LISTEN, STOL_TCPSTATE (-3),
          STOL_TCPSTATE (-3) == LX_TCP_LISTEN ? "OK" : "ERR");
/*
  printf ("TCPS_BOUND %d LX_TCP_BOUND %d result %d---> %s \n",
          TCPS_BOUND, LX_TCP_BOUND, STOL_TCPSTATE (-4),
          STOL_TCPSTATE (-4) == LX_TCP_BOUND ? "OK" : "ERR");

  printf ("TCPS_IDLE %d LX_TCP_IDLE %d result %d---> %s \n",
          TCPS_IDLE, LX_TCP_IDLE, STOL_TCPSTATE (-5),
          STOL_TCPSTATE (-5) == LX_TCP_IDLE ? "OK" : "ERR");
*/
  printf ("TCPS_CLOSED %d LX_TCP_CLOSE %d result %d---> %s \n",
          TCPS_CLOSED, LX_TCP_CLOSE, STOL_TCPSTATE (-6),
          STOL_TCPSTATE (-6) == LX_TCP_CLOSE ? "OK" : "ERR");

  printf ("TCPS_BOUND %d should return LX_TCP_CLOSE_WAIT %d result %d---> %s \n",
          TCPS_BOUND, LX_TCP_CLOSE_WAIT, STOL_TCPSTATE (-4),
          STOL_TCPSTATE (-4) == LX_TCP_CLOSE_WAIT ? "OK" : "ERR");

}
```
I tested  this:
```c
s = socket(AF_INET, SOCK_STREAM, 0);
```
In a Linux vm and a lx branded zone with this change returns
```
connect() failed: Address family not supported by protocol
```